### DISCUSSION TITLE
voodoo - a 3D game programming library

### DISCUSSION THREAD

**Discussion by zacharycarter** [janet-lang/janet#1353] [Show and tell]
Hello!

I've been itching to work on a project which serves as an excuse to try out Janet. This is what I've created in an attempt to scratch that itch: [voodoo](https://github.com/zacharycarter/voodoo),

It's far from complete or even useful really. I was previously working on a browser game, which I intend to use as dogfood for this project.

I'll try to get a hosted version online soon.

Shout out to @ianthehenry for creating Bauble studio as it's' been an invaluable source of ideas and example code. Also for the book and blog posts he's written about Janet. Also thank you @sogaiu for answering questions for me in the Janet Zulip help channel, and everyone that has and currently contributes to Janet. I've found it to be a very enjoyable language to code in.

As I make more progress I'll post updates here.

**Comment by AlecTroemel** [comment]
this is really cool! I've had a similar idea bouncing around in my head for a while. What does your roadmap look like for Voodoo? 

Any thoughts on supporting 2D? 

Being able to open up multiple textareas, similar to [this clojure project](https://github.com/adam-james-v/badspreadsheet), would be _very cool_

**Reply by zacharycarter** [reply]
Hi @AlecTroemel! 

Thank you for the reply and the kind words! I've been kind of winging the roadmap, but it might be a good idea to formalize one so that I have a plan to stick to, not to mention being able to share it with anyone interested. I'll try to find some time this week to throw one together.

Currently I'm working on implementing a virtual file system and code for managing assets. I want to attempt to keep the API relatively high level, so this is all pretty boring plumbing work to enable them.

I do plan on supporting 2D eventually and that Clojure project does look neat!

I do plan on allowing scripts to be attached to objects in a scene and plans for some sort of visual scripting - I'm not sure I have ambitions at the moment for allowing the user to open text editors like in that screenshot. I'm planning to expose the virtual file system to the user so they can open different scripts and edit them, but that's a bit different than what you're describing.

The editor in the screenshot isn't hardware accelerated - it's simply an instance of [CodeMirror](https://codemirror.net/) very similar to how [bauble.studio](https://bauble.studio/) works. With the GUI API I eventually have planned for the library, I imagine it would be possible to create something to suit this requirement.

Thanks again for the questions and feedback!

**Reply by zacharycarter** [reply]
Hi @AlecTroemel (and anyone else who feels like following along :wave)!

I'm sorry it took me a week longer than anticipated, but I managed to get a very rough roadmap added to the readme. I'm currently searching for work while simultaneously taking care of a pregnant wife and an eighteen-month-old, so to say that my life is busy at the moment, at least for me anyway, is a bit of an understatement 😅.

I've still managed to make a bit of progress on the project since I typed up my last reply. Besides the updates to the readme, I managed to get pthreads and fibers working in the browser and was able to get a fiber-based job system implemented and working. I used address sanitizer to help verify that I wasn't destroying every stack I encountered and it seems to be happy. I'm using this job system right now for loading assets from the virtual file system I implemented on top of Emscripten's IDBFS, but theoretically it could be used to accomplish any short-lived I/O or cpu bound workloads. Also, re: the virtual file system - the plan there is to be able to bundle scripts and assets as modules or mods and then persist any changes to them as well. I haven't fully worked this part out yet, but theoretically it's all possible and just requires me to do the work.

I'm working towards getting something more impressive than a cube drawn, so I've started the trudge towards implementing skeletal animation and at the same time building out the forward renderer. Hopefully by the time I reply next, I'll have a decent forward renderer with shadows, atmospheric effects, fxaa, AO and some other features implemented, along with skeletal animation. Once that's done I'll start working on other things needed to build an actual game using the project - like an animation and player controller and things like that.

Any who, I just wanted to let you know that I hadn't forgotten about your kind words and do very much appreciate you taking interest in the project. I also started back up a website for myself - https://tailwag.games/ - the first blog post is pretty terrible but was also just kind of a late-night thought dump and probably is only worth reading if you know me personally. Having said that, I'm planning a series of technical deep dives into voodoo. I've already finished drafting one that goes into how the scripting subsystem works.

As soon as I have the updates to the renderer implemented and skeletal animation working, I'm planning on hosting a version on my website so folks can check it out and play around with it.  I'll reply here once I do, along with any blog posts I drop along the way.

Thanks again for the kind reply, and I hope some of the above was of interest to you! If you have any questions or comments, I'd love to hear them!

**Reply by pepe** [reply]
OT. I wish your wife the best of luck, and please take as much care as possible so she stays in the most pleasing mood. And say hi to your older one 👋🏾.

**Reply by pepe** [reply]
PS: I have read your blog post, and oh man, I wish I could write half as well as you do. "From the bottom of my heart" :-).

**Reply by AlecTroemel** [reply]
@zacharycarter as someone with a 2 year old and another on the way, I can relate to time for projects being hard to come by :smile: 

I also ready through your blog, and agree with pepe that it was well written! I'm want to read more from you (does your site have an rss feed? I couldnt find one poking around)

Thank you for the roadmap in voodoo! Reading through it, there may be a couple places I can contribute. Also, If I can swing it, I may try using voodoo in a gamejam this spring. I have very little experience in 3d, so it would be a learning experience for sure.

**Reply by zacharycarter** [reply]
@Pepe and @AlecTroemel please accept my sincerest apologies for not replying to your comments sooner! I've been a bit swamped with job interviews, and life in general. I'm starting to get into the technical rounds, so hopefully my job search will be wrapping up in the next couple, or few weeks. 

I very much appreciate the kind words from both of you regarding my blog. I noticed a bunch of spelling and grammatical errors after reading it, and it was a bit all over the place, but I think my next posts will be improvements. I hope so anyway! @AlecTroemel I will work on getting RSS support added when I publish the upcoming entries - thanks for that feedback, It's a wonderful suggestion!

When I'm back to working on the project again more regularly, hopefully during the next month or two, I'd love to talk more about potential contributions! I recently added - https://github.com/SanderMertens/flecs - into the mix. I had started on a Janet native module for the library, and then switched gears to focus on lower-level tasks. I think it's a good time to introduce this type of functionality into the project.

I really don't like the idea of forcing users to rely on an ECS, so I hope I can find some sort of abstraction to remove some of the pain around using them. My goal is to provide a bunch of default features that the user can replace with their own if desired. Flecs has a `meta` module which allows one to define components at runtime and generates reflection data for them. This should eventually aid in the building of the editor. I think I can probably get away with some sort of spatial hierarchy using flecs, and expose an interface that feels similar to an object component design.

I'm diving down the rabbit hole again, so I'll end it here 😆. Thanks once again for the kind words from both of you! I hope you both are having amazing Februarys! I'll post another comment when the blog posts are up. Take care!

**Reply by zacharycarter** [reply]
Just a quick update - I recently added shadows and skeletal animation to the project. I plan on working on input and a player controller next. After I finish with those tasks, I'll most likely take a pause and write some more blog entries and work on getting a live version hosted. I'll post an update here when I do so.
